realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -1128,23 +1169,22 @@ public void run() {
                     }
 
                     // Send response as the final step to ensure the bg thread quit before others get the response!
-                    if (callback != null
-                            && handler != null
+                    if (handler != null
                             && !Thread.currentThread().isInterrupted()
                             && handler.getLooper().getThread().isAlive()) {
-                        if (transactionCommitted) {
+                        if (onSuccess != null && transactionCommitted) {
                             handler.post(new Runnable() {
                                 @Override
                                 public void run() {
-                                    callback.onSuccess();
+                                    onSuccess.onSuccess();
                                 }
                             });
-                        } else if (exception[0] != null) {
-                            // transaction has not been canceled by there is a exception during transaction.
+                        }
+                        if (onError != null && exception[0] != null) {
```

Comment:
If there is an exception, and user doesn't supply a `onError`, I think we should just throw.